{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test Cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These genes and primers are compatible\n",
    "gene = \"atggagacagacacactcctgctatgggtactgctgctctgggttccaggttccactggtgacacaagtttgtacaaaaaagttggcaccaagtcgatcctagatggccttgcagataccaccttccgcaccatcaccactgacctcctgtacgtgggctcaaatgacattcagtacgaagacatcaaaggtgacatggcatccaaattagggtacttcccacagaaattccctttaacttcctttaggggaagtcccttccaagagaagatgactgcgggagacaacccccagctagtcccagcagaccaggtgaacattacagaattttacaacaagtctctctcgtccttcaaggagaatgaggagaacatccagtgtggggagaacttcatggacatagagtgtttcatggtcctgaaccccagccagcagctggccattgcagtcctgtccctcacgctgggcaccttcacggtcctggagaacctcctggtgctgtgcgtcatcctccactcccgcagcctccgctgcaggccttcctaccacttcatcggcagcctggcggtggcagacctcctggggagtgtcatttttgtctacagcttcattgacttccacgtgttccaccgcaaagatagccgcaacgtgtttctgttcaaactgggtggggtcacggcctccttcactgcctccgtgggcagcctgttcctcacagccatcgacaggtacatatccattcacaggcccctggcctataagaggattgtcaccaggcccaaggccgtggtggcgttttgcctgatgtggaccatagccattgtgatcgccgtgctgcctctcctgggctggaactgcgagaaactgcaatctgtttgctcagacattttcccacacattgatgaaacctacctgatgttctggatcggggtcaccagcgtactgcttctgttcatcgtgtatgcgtacatgtatattctctggaaggctcacagccacgccgtccgcatgattcagcgtaccgacgcgctggacctggaggagggaggaaacgtctatatcaaggccgacaagcagaagaacggcatcaaggcgaacttctgcatccgccacaacatcgaggacggcggcgtgcagctcgcctaccactaccagcagaacacccccatcggcgacggccccgtgctgctgcccgacaaccactacctgagcgtgcagtccaaactttcgaaagaccccaacgagaagcgcgatcacatggtcctgctggagttcgtgaccgccgccgggatcactttcggcatggacgagctgtacaagggcggtaccggagggagcatggtgagaaagggcgaggagctgttcaccggggtggtgcccatcctggtcgagctggacggcgacgtaaacggccacaagttcagcgtgggcggcgagggtgagggcgatgccaccgttggcaagctgaccctgaagttcatctgcaccaccggcaagctgcccgtgccctggcccaccctcgtgaccaccctgacctacggcgtgcagtgcttcagccgctaccccgaccacatgaagcagcacgacttcttcaagtccgccatgcccgaaggctacatccaggagcgcaccatcttcttcaaggacgacggcaactacaagacccgcgccgaggtgaagttcgagggcgacaccctggtgaaccgcatcgagctgaagggcatcgacttcaaggaggacggcaacatcctggggcacaagctggagtacaacaccggagcagcagcacgctggcgcgggcggcgcatggacattaggttagccaagaccctggtcctgatcctggtggtgttgatcatctgctggggccctctgcttgcaatcatggtgtatgatgtctttgggaagatgaacaagctcattaagacggtgtttgcattctgcaccatgctctgcctgctgaactccaccgtgaaccccatcatctatgctctgaggagtaaggacctgcgacacgctttccggagcatgtttccctcttgtgaaggcactgcgcagcctctggataacagcatgggggactcggactgcctgcacaaacacgcaaacaatgcagccagtgttcacagggccgcagaaagctgcatcaagagcacggtcaagattgccaaggtaaccatgtctgtgtccacagacacgtctgccgaggctctg\"\n",
    "forward_primer = \"atggagacagacacactcctgctatgg\"\n",
    "reverse_primer = \"cagagcctcggcagacgtgt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sourced from internet | Gene: \"https://www.bioinformatics.org/sms2/random_dna.html\" , Forward primer & Reverse Primer: \"https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi\"\n",
    "gene1 = \"acggagatttcccccttcctagttccctaatggataagatgtttaagatgttcaaacaaggatattcactgtgccagccctgtggtcgggaagtcgaata\"\n",
    "forward_primer1 = \"ACGGAGATTTCCCCCTTCCTA\" # binds to complement templete | -> direction\n",
    "reverse_primer1 = \"TATTCGACTTCCCGACCACAG\" # binds to original templete | reverse primer then <- direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sourced from internet positions are different \n",
    "gene0 = \"caatatggtgagcaatcttttgctcttggggattgaccacagtcgtcgacgcgatgcttgccctagtaagcttggcgtacaattacaagtgataagaggt\"\n",
    "forward_primer0 = \"GAGCAATCTTTTGCTCTTGGGG\" \n",
    "reverse_primer0 = \"TACGCCAAGCTTACTAGGGC\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Completed Funtions Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK!\n"
     ]
    }
   ],
   "source": [
    "class pcr(object): \n",
    "  def __init__ (self, gene, forward_primer, reverse_primer, template_type = \"plasmid\", enzymes = [], factor = \"cost\"): \n",
    "    self.sequence = gene\n",
    "    self.length = len(gene)\n",
    "    self.factor = factor\n",
    "    self.template_type = template_type\n",
    "    self.fp = forward_primer\n",
    "    self.rp = reverse_primer\n",
    "\n",
    "# Tests\n",
    "test = pcr(\"agctgctgaaacccgggtttaatcgggddd\", \"atggagacagacacactcctgctatgg\", \"acacgtctgccgaggctctg\", \"genomic\")  # made changes to the init code and made plasmids defult \n",
    "assert(test.fp == \"atggagacagacacactcctgctatgg\")\n",
    "assert(test.rp == \"acacgtctgccgaggctctg\")\n",
    "assert(test.length == len(\"agctgctgaaacccgggtttaatcgggddd\"))\n",
    "print(\"OK!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "def countDNAbase(gene): \n",
    "  \"\"\"\n",
    "  Counts the number of each DNA base present in a sequence\n",
    "\n",
    "  Parameters\n",
    "  ----------\n",
    "  dna: input gene \n",
    "\n",
    "  Returns\n",
    "  ----------\n",
    "  The number of each base \n",
    "  GC content : str\n",
    "\n",
    "  \"\"\"\n",
    "  a_count = 0\n",
    "  c_count = 0\n",
    "  g_count = 0\n",
    "  t_count = 0\n",
    "  for char in gene: \n",
    "    if char == \"a\": \n",
    "      a_count += 1 \n",
    "    if char == \"c\": \n",
    "      c_count += 1 \n",
    "    if char == \"g\": \n",
    "      g_count += 1 \n",
    "    if char == \"t\": \n",
    "      t_count += 1\n",
    "  gc_content = 100*(g_count + c_count)/(g_count + c_count + a_count + t_count)\n",
    "  return \"The GC Content is\",gc_content, \"%\"\n",
    "\n",
    "# Tests\n",
    "assert(countDNAbase(\"aagc\") == ('The GC Content is', 50.0, '%'))\n",
    "print(\"OK\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "def checkPrimerGeneCompatability(gene, forward_primer, reverse_primer, startr = 0, stopr = 0 , startf = 0, stopf = 0): \n",
    "  \"\"\"\n",
    "  Check the forward and reverse primer's compatibility to the gene/sequence.\n",
    "  If the start positions are not at the begining on the gene you can enter the \n",
    "  start and stop positions in acordance with Primer-blast\n",
    "  \n",
    "  Parameters\n",
    "  ----------\n",
    "  gene: str\n",
    "  forward_primer: str \n",
    "  reverse_primer: str\n",
    "  startr: int\n",
    "  stopr: int\n",
    "  startf: int\n",
    "  stopf: int\n",
    "  \n",
    "  Returns\n",
    "  -------\n",
    "  A statment on primer compatibility : str\n",
    "  \"\"\"\n",
    "  # Preparing inputs \n",
    "  gene = gene.lower()\n",
    "  forward_primer = forward_primer.lower()\n",
    "  reverse_primer = reverse_primer.lower()\n",
    "\n",
    "  # Create complement squence for comparison with forward_primer\n",
    "  complement = gene.upper().replace(\"A\", \"t\").replace(\"C\", \"g\").replace(\"T\", \"a\").replace(\"G\", \"c\")\n",
    "\n",
    "  # Locating the start and stop pointsoption\n",
    "  if startr == 0 and stopr == 0:\n",
    "    gene_reverse = gene[-len(reverse_primer):]\n",
    "  else:\n",
    "    gene_reverse = gene[stopr-1:startr]\n",
    "  if startf == 0 and stopf == 0:\n",
    "    gene_forward = complement[:len(forward_primer)]\n",
    "  else:\n",
    "    gene_forward = complement[startf-1:stopf]\n",
    "\n",
    "  # Create gene comparison points\n",
    "  gene_reverse = gene_reverse[::-1] # Reverse Primers are entered in Reverse so this is ment to flip the gene to compare\n",
    "\n",
    "  # Ensure gene and primer compatibility\n",
    "  position_gene = 0\n",
    "  Primer1 = 0\n",
    "  Primer2 = 0\n",
    "  for base_gene in gene_forward:\n",
    "    if base_gene == \"a\" and forward_primer[position_gene] == \"t\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    elif base_gene == \"t\" and forward_primer[position_gene] == \"a\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    elif base_gene == \"c\" and forward_primer[position_gene] == \"g\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    elif base_gene == \"g\" and forward_primer[position_gene] == \"c\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    else: \n",
    "      Primer1 = 1\n",
    "      break\n",
    "      \n",
    "  position_gene = 0\n",
    "  for base_gene in gene_reverse:\n",
    "    if base_gene == \"a\" and reverse_primer[position_gene] == \"t\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    elif base_gene == \"t\" and reverse_primer[position_gene] == \"a\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    elif base_gene == \"c\" and reverse_primer[position_gene] == \"g\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    elif base_gene == \"g\" and reverse_primer[position_gene] == \"c\":\n",
    "      position_gene += 1\n",
    "      continue\n",
    "    else: \n",
    "      Primer2 = 1\n",
    "      break\n",
    "      \n",
    "  if Primer1 + Primer2 == 2:\n",
    "    return (\"Both Primers and gene are incompatible\")\n",
    "  elif Primer1 == 1:\n",
    "    return (\"Forward Primer and gene are incompatible\")   \n",
    "  elif Primer2 == 1:\n",
    "    return (\"Reverse Primer and gene are incompatible\")\n",
    "  else:\n",
    "    return (\"Primers and Gene are compatible\")\n",
    "\n",
    "# Test\n",
    "assert(checkPrimerGeneCompatability(gene, forward_primer, reverse_primer) == \"Primers and Gene are compatible\")\n",
    "assert(checkPrimerGeneCompatability(gene1, forward_primer1, reverse_primer1) == \"Primers and Gene are compatible\")\n",
    "assert(checkPrimerGeneCompatability(gene0, forward_primer0, reverse_primer0, startr = 79, stopr = 60 , startf = 10, stopf = 31) == \"Primers and Gene are compatible\")\n",
    "print(\"OK\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OK\n"
     ]
    }
   ],
   "source": [
    "#This one uses Modified Bresluaer Values for Phusion \n",
    "#Paper: https://academic.oup.com/bioinformatics/article/21/6/711/199347\n",
    "def meltingTemperaturePhusion(seq): \n",
    "  \"\"\"\n",
    "  checks primer melting temperature (Tm)\n",
    "  Adapted from Sigma Aldrich protocol: https://www.sigmaaldrich.com/US/en/technical-documents/protocol/genomics/pcr/oligos-melting-temp\n",
    "  Santa Lucia calculations\n",
    "  Tm = temp where 50% of DNA is melted, and 50% of DNA is double stranded\n",
    "\n",
    "  Parameters: \n",
    "  ------------\n",
    "  seq : Primer sequnce as a string\n",
    "\n",
    "  Returns: \n",
    "  ------------\n",
    "  tm : int\n",
    "    DNA melting temperature in degrees celsius\n",
    "\n",
    "  \"\"\"\n",
    "  import math as math \n",
    "  H = 0 \n",
    "  S = 0 \n",
    "\n",
    "  for x in range(0,len(seq)+1): \n",
    "    check = str(seq[x:x+2]).lower()\n",
    " \n",
    "    if check == \"aa\" or check  == \"tt\": \n",
    "      H += -9.1\n",
    "      S += -0.0240\n",
    "\n",
    "    elif check == \"at\": \n",
    "      H += -8.6\n",
    "      S += -0.0239\n",
    "\n",
    "    elif check == \"ta\": \n",
    "      H += -6.0\n",
    "      S += -0.0169\n",
    "\n",
    "    elif check == \"ca\" or check == \"tg\": \n",
    "      H += -5.8\n",
    "      S += -0.0129\n",
    "\n",
    "    elif check == \"gt\" or check == \"ac\": \n",
    "      H += -6.5 \n",
    "      S += -0.0173\n",
    "\n",
    "    elif check == \"ct\" or check == \"ag\": \n",
    "      H += -7.8 \n",
    "      S += -0.0208\n",
    "\n",
    "    elif check == \"ga\" or check == \"tc\": \n",
    "      H += -5.6 \n",
    "      S += -0.0135\n",
    "\n",
    "    elif check == \"cg\": \n",
    "      H += -11.9\n",
    "      S += -0.0278\n",
    "\n",
    "    elif check == \"gc\": \n",
    "      H += -11.1\n",
    "      S += -0.0267\n",
    "\n",
    "    elif check == \"gg\" or check == \"cc\": \n",
    "      H += -11.0\n",
    "      S += -0.0266\n",
    "\n",
    "  tm = (H/(-0.0108+S+0.00199*math.log(0.0000005/4)))-273.15+16.6*math.log10(0.05)\n",
    "  return tm\n",
    "\n",
    "#Test: \n",
    "assert(float(meltingTemperaturePhusion(\"atggagacagacacactcctgctatgg\")))\n",
    "print(\"OK\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Class Creation Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "class pcr(object): \n",
    "  def __init__ (self, gene, forward_primer, reverse_primer, template_type = \"plasmid\", enzymes = [], factor = \"cost\"): \n",
    "    self.sequence = gene\n",
    "    self.length = len(gene)\n",
    "    self.factor = factor\n",
    "    self.template_type = template_type\n",
    "    self.fp = forward_primer\n",
    "    self.rp = reverse_primer\n",
    "\n",
    "  def countDNAbase(gene): \n",
    "    \"\"\"\n",
    "    Counts the number of each DNA base present in a sequence\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dna: input gene \n",
    "\n",
    "    Returns\n",
    "    ----------\n",
    "    The number of each base \n",
    "    GC content : str\n",
    "\n",
    "    \"\"\"\n",
    "    a_count = 0\n",
    "    c_count = 0\n",
    "    g_count = 0\n",
    "    t_count = 0\n",
    "    for char in gene: \n",
    "      if char == \"a\": \n",
    "        a_count += 1 \n",
    "      if char == \"c\": \n",
    "        c_count += 1 \n",
    "      if char == \"g\": \n",
    "        g_count += 1 \n",
    "      if char == \"t\": \n",
    "        t_count += 1\n",
    "    gc_content = 100*(g_count + c_count)/(g_count + c_count + a_count + t_count)\n",
    "    return \"The GC Content is\",gc_content, \"%\"\n",
    "\n",
    "  def checkPrimerGeneCompatability(gene, forward_primer, reverse_primer, startr = 0, stopr = 0 , startf = 0, stopf = 0): \n",
    "    \"\"\"\n",
    "    Check the forward and reverse primer's compatibility to the gene/sequence.\n",
    "    If the start positions are not at the begining on the gene you can enter the \n",
    "    start and stop positions in acordance with Primer-blast\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    gene: str\n",
    "    forward_primer: str \n",
    "    reverse_primer: str\n",
    "    startr: int\n",
    "    stopr: int\n",
    "    startf: int\n",
    "    stopf: int\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    A statment on primer compatibility : str\n",
    "    \"\"\"\n",
    "    # Preparing inputs \n",
    "    gene = gene.lower()\n",
    "    forward_primer = forward_primer.lower()\n",
    "    reverse_primer = reverse_primer.lower()\n",
    "\n",
    "    # Create complement squence for comparison with forward_primer\n",
    "    complement = gene.upper().replace(\"A\", \"t\").replace(\"C\", \"g\").replace(\"T\", \"a\").replace(\"G\", \"c\")\n",
    "\n",
    "    # Locating the start and stop pointsoption\n",
    "    if startr == 0 and stopr == 0:\n",
    "      gene_reverse = gene[-len(reverse_primer):]\n",
    "    else:\n",
    "      gene_reverse = gene[stopr-1:startr]\n",
    "    if startf == 0 and stopf == 0:\n",
    "      gene_forward = complement[:len(forward_primer)]\n",
    "    else:\n",
    "      gene_forward = complement[startf-1:stopf]\n",
    "\n",
    "    # Create gene comparison points\n",
    "    gene_reverse = gene_reverse[::-1] # Reverse Primers are entered in Reverse so this is ment to flip the gene to compare\n",
    "\n",
    "    # Ensure gene and primer compatibility\n",
    "    position_gene = 0\n",
    "    Primer1 = 0\n",
    "    Primer2 = 0\n",
    "    for base_gene in gene_forward:\n",
    "      if base_gene == \"a\" and forward_primer[position_gene] == \"t\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      elif base_gene == \"t\" and forward_primer[position_gene] == \"a\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      elif base_gene == \"c\" and forward_primer[position_gene] == \"g\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      elif base_gene == \"g\" and forward_primer[position_gene] == \"c\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      else: \n",
    "        Primer1 = 1\n",
    "        break\n",
    "        \n",
    "    position_gene = 0\n",
    "    for base_gene in gene_reverse:\n",
    "      if base_gene == \"a\" and reverse_primer[position_gene] == \"t\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      elif base_gene == \"t\" and reverse_primer[position_gene] == \"a\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      elif base_gene == \"c\" and reverse_primer[position_gene] == \"g\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      elif base_gene == \"g\" and reverse_primer[position_gene] == \"c\":\n",
    "        position_gene += 1\n",
    "        continue\n",
    "      else: \n",
    "        Primer2 = 1\n",
    "        break\n",
    "        \n",
    "    if Primer1 + Primer2 == 2:\n",
    "      return (\"Both Primers and gene are incompatible\")\n",
    "    elif Primer1 == 1:\n",
    "      return (\"Forward Primer and gene are incompatible\")   \n",
    "    elif Primer2 == 1:\n",
    "      return (\"Reverse Primer and gene are incompatible\")\n",
    "    else:\n",
    "      return (\"Primers and Gene are compatible\")\n",
    "\n",
    "  def meltingTemperaturePhusion(seq): \n",
    "    \"\"\"\n",
    "    checks primer melting temperature (Tm)\n",
    "    Adapted from Sigma Aldrich protocol: https://www.sigmaaldrich.com/US/en/technical-documents/protocol/genomics/pcr/oligos-melting-temp\n",
    "    Santa Lucia calculations\n",
    "    Tm = temp where 50% of DNA is melted, and 50% of DNA is double stranded\n",
    "\n",
    "    Parameters: \n",
    "    ------------\n",
    "    seq : Primer sequnce as a string\n",
    "\n",
    "    Returns: \n",
    "    ------------\n",
    "    tm : int\n",
    "      DNA melting temperature in degrees celsius\n",
    "\n",
    "    \"\"\"\n",
    "    import math as math \n",
    "    H = 0 \n",
    "    S = 0 \n",
    "\n",
    "    for x in range(0,len(seq)+1): \n",
    "      check = str(seq[x:x+2]).lower()\n",
    "  \n",
    "      if check == \"aa\" or check  == \"tt\": \n",
    "        H += -9.1\n",
    "        S += -0.0240\n",
    "\n",
    "      elif check == \"at\": \n",
    "        H += -8.6\n",
    "        S += -0.0239\n",
    "\n",
    "      elif check == \"ta\": \n",
    "        H += -6.0\n",
    "        S += -0.0169\n",
    "\n",
    "      elif check == \"ca\" or check == \"tg\": \n",
    "        H += -5.8\n",
    "        S += -0.0129\n",
    "\n",
    "      elif check == \"gt\" or check == \"ac\": \n",
    "        H += -6.5 \n",
    "        S += -0.0173\n",
    "\n",
    "      elif check == \"ct\" or check == \"ag\": \n",
    "        H += -7.8 \n",
    "        S += -0.0208\n",
    "\n",
    "      elif check == \"ga\" or check == \"tc\": \n",
    "        H += -5.6 \n",
    "        S += -0.0135\n",
    "\n",
    "      elif check == \"cg\": \n",
    "        H += -11.9\n",
    "        S += -0.0278\n",
    "\n",
    "      elif check == \"gc\": \n",
    "        H += -11.1\n",
    "        S += -0.0267\n",
    "\n",
    "      elif check == \"gg\" or check == \"cc\": \n",
    "        H += -11.0\n",
    "        S += -0.0266\n",
    "\n",
    "    tm = (H/(-0.0108+S+0.00199*math.log(0.0000005/4)))-273.15+16.6*math.log10(0.05)\n",
    "    return tm\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 64-bit (microsoft store)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "55b4ec3b1de4745d4d0bc52a98f917eca26cf669fe460c5ea602bd5130f820d0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
