{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test Cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# These genes and primers are compatible\n",
    "gene = \"atggagacagacacactcctgctatgggtactgctgctctgggttccaggttccactggtgacacaagtttgtacaaaaaagttggcaccaagtcgatcctagatggccttgcagataccaccttccgcaccatcaccactgacctcctgtacgtgggctcaaatgacattcagtacgaagacatcaaaggtgacatggcatccaaattagggtacttcccacagaaattccctttaacttcctttaggggaagtcccttccaagagaagatgactgcgggagacaacccccagctagtcccagcagaccaggtgaacattacagaattttacaacaagtctctctcgtccttcaaggagaatgaggagaacatccagtgtggggagaacttcatggacatagagtgtttcatggtcctgaaccccagccagcagctggccattgcagtcctgtccctcacgctgggcaccttcacggtcctggagaacctcctggtgctgtgcgtcatcctccactcccgcagcctccgctgcaggccttcctaccacttcatcggcagcctggcggtggcagacctcctggggagtgtcatttttgtctacagcttcattgacttccacgtgttccaccgcaaagatagccgcaacgtgtttctgttcaaactgggtggggtcacggcctccttcactgcctccgtgggcagcctgttcctcacagccatcgacaggtacatatccattcacaggcccctggcctataagaggattgtcaccaggcccaaggccgtggtggcgttttgcctgatgtggaccatagccattgtgatcgccgtgctgcctctcctgggctggaactgcgagaaactgcaatctgtttgctcagacattttcccacacattgatgaaacctacctgatgttctggatcggggtcaccagcgtactgcttctgttcatcgtgtatgcgtacatgtatattctctggaaggctcacagccacgccgtccgcatgattcagcgtaccgacgcgctggacctggaggagggaggaaacgtctatatcaaggccgacaagcagaagaacggcatcaaggcgaacttctgcatccgccacaacatcgaggacggcggcgtgcagctcgcctaccactaccagcagaacacccccatcggcgacggccccgtgctgctgcccgacaaccactacctgagcgtgcagtccaaactttcgaaagaccccaacgagaagcgcgatcacatggtcctgctggagttcgtgaccgccgccgggatcactttcggcatggacgagctgtacaagggcggtaccggagggagcatggtgagaaagggcgaggagctgttcaccggggtggtgcccatcctggtcgagctggacggcgacgtaaacggccacaagttcagcgtgggcggcgagggtgagggcgatgccaccgttggcaagctgaccctgaagttcatctgcaccaccggcaagctgcccgtgccctggcccaccctcgtgaccaccctgacctacggcgtgcagtgcttcagccgctaccccgaccacatgaagcagcacgacttcttcaagtccgccatgcccgaaggctacatccaggagcgcaccatcttcttcaaggacgacggcaactacaagacccgcgccgaggtgaagttcgagggcgacaccctggtgaaccgcatcgagctgaagggcatcgacttcaaggaggacggcaacatcctggggcacaagctggagtacaacaccggagcagcagcacgctggcgcgggcggcgcatggacattaggttagccaagaccctggtcctgatcctggtggtgttgatcatctgctggggccctctgcttgcaatcatggtgtatgatgtctttgggaagatgaacaagctcattaagacggtgtttgcattctgcaccatgctctgcctgctgaactccaccgtgaaccccatcatctatgctctgaggagtaaggacctgcgacacgctttccggagcatgtttccctcttgtgaaggcactgcgcagcctctggataacagcatgggggactcggactgcctgcacaaacacgcaaacaatgcagccagtgttcacagggccgcagaaagctgcatcaagagcacggtcaagattgccaaggtaaccatgtctgtgtccacagacacgtctgccgaggctctg\"\n",
    "forward_primer = \"atggagacagacacactcctgctatgg\"\n",
    "reverse_primer = \"cagagcctcggcagacgtgt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sourced from internet | Gene: \"https://www.bioinformatics.org/sms2/random_dna.html\" , Forward primer & Reverse Primer: \"https://www.ncbi.nlm.nih.gov/tools/primer-blast/index.cgi\"\n",
    "gene1 = \"acggagatttcccccttcctagttccctaatggataagatgtttaagatgttcaaacaaggatattcactgtgccagccctgtggtcgggaagtcgaata\"\n",
    "forward_primer1 = \"ACGGAGATTTCCCCCTTCCTA\" # binds to complement templete | -> direction\n",
    "reverse_primer1 = \"TATTCGACTTCCCGACCACAG\" # binds to original templete | reverse primer then <- direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sourced from internet positions are different \n",
    "gene0 = \"caatatggtgagcaatcttttgctcttggggattgaccacagtcgtcgacgcgatgcttgccctagtaagcttggcgtacaattacaagtgataagaggt\"\n",
    "forward_primer0 = \"GAGCAATCTTTTGCTCTTGGGG\" \n",
    "reverse_primer0 = \"TACGCCAAGCTTACTAGGGC\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Completed Class testing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Class Creation Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Requirements\n",
    "import pandas as pd\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class pcr(object): \n",
    "    def __init__ (self, gene, forward_primer, reverse_primer, template_type = \"plasmid\"):\n",
    "        \"\"\"\n",
    "        gene: target gene being amplified\n",
    "        forward_primer: primer used for forward reaction ***should only be portion of primer that overlaps to gene***\n",
    "        reverse_primer: primer used for reverse reaction ***should only be portion of primer that overlaps to gene***\n",
    "        template_type: type of DNA being amplified, either \"plasmid\" or \"genomic\n",
    "\n",
    "        \"\"\"\n",
    "        self.sequence = gene\n",
    "        self.length = len(gene)\n",
    "        self.fp = forward_primer\n",
    "        self.rp = reverse_primer\n",
    "        self.template_type = template_type\n",
    "      \n",
    "    def countDNAbase(self): \n",
    "        \"\"\"\n",
    "        Counts the number of each DNA base present in a sequence\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        dna: input gene \n",
    "\n",
    "        Returns\n",
    "        ----------\n",
    "        The number of each base \n",
    "        GC content : str\n",
    "\n",
    "        \"\"\"\n",
    "        gene = self.sequence \n",
    "        \n",
    "        a_count = 0\n",
    "        c_count = 0\n",
    "        g_count = 0\n",
    "        t_count = 0\n",
    "        for char in gene: \n",
    "          if char == \"a\": \n",
    "            a_count += 1 \n",
    "          if char == \"c\": \n",
    "            c_count += 1 \n",
    "          if char == \"g\": \n",
    "            g_count += 1 \n",
    "          if char == \"t\": \n",
    "            t_count += 1\n",
    "        gc_content = 100*(g_count + c_count)/(g_count + c_count + a_count + t_count)\n",
    "        return \"The GC Content is {}\".format(gc_content), gc_content\n",
    "\n",
    "    def checkPrimerGeneCompatability(self, startr = 0, stopr = 0 , startf = 0, stopf = 0): \n",
    "        \"\"\"\n",
    "        Check the forward and reverse primer's compatibility to the gene/sequence.\n",
    "        If the start positions are not at the begining on the gene you can enter the \n",
    "        start and stop positions in acordance with Primer-blast\n",
    "        \n",
    "        Parameters\n",
    "        ----------\n",
    "        gene: str\n",
    "        forward_primer: str \n",
    "        reverse_primer: str\n",
    "        startr: int\n",
    "        stopr: int\n",
    "        startf: int\n",
    "        stopf: int\n",
    "        \n",
    "        Returns\n",
    "        -------\n",
    "        A statment on primer compatibility : str\n",
    "        \"\"\"\n",
    "        \n",
    "        forward_primer = self.fp\n",
    "        reverse_primer = self.rp\n",
    "        gene = self.sequence\n",
    "\n",
    "        # Preparing inputs \n",
    "        gene = gene.lower()\n",
    "        forward_primer = forward_primer.lower()\n",
    "        reverse_primer = reverse_primer.lower()\n",
    "\n",
    "        # Create complement squence for comparison with forward_primer\n",
    "        complement = gene.upper().replace(\"A\", \"t\").replace(\"C\", \"g\").replace(\"T\", \"a\").replace(\"G\", \"c\")\n",
    "\n",
    "        # Locating the start and stop pointsoption\n",
    "        if startr == 0 and stopr == 0:\n",
    "          gene_reverse = gene[-len(reverse_primer):]\n",
    "        else:\n",
    "          gene_reverse = gene[stopr-1:startr]\n",
    "        if startf == 0 and stopf == 0:\n",
    "          gene_forward = complement[:len(forward_primer)]\n",
    "        else:\n",
    "          gene_forward = complement[startf-1:stopf]\n",
    "\n",
    "        # Create gene comparison points\n",
    "        gene_reverse = gene_reverse[::-1] # Reverse Primers are entered in Reverse so this is ment to flip the gene to compare\n",
    "\n",
    "        # Ensure gene and primer compatibility\n",
    "        position_gene = 0\n",
    "        Primer1 = 0\n",
    "        Primer2 = 0\n",
    "        for base_gene in gene_forward:\n",
    "          if base_gene == \"a\" and forward_primer[position_gene] == \"t\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          elif base_gene == \"t\" and forward_primer[position_gene] == \"a\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          elif base_gene == \"c\" and forward_primer[position_gene] == \"g\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          elif base_gene == \"g\" and forward_primer[position_gene] == \"c\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          else: \n",
    "            Primer1 = 1\n",
    "            break\n",
    "            \n",
    "        position_gene = 0\n",
    "        for base_gene in gene_reverse:\n",
    "          if base_gene == \"a\" and reverse_primer[position_gene] == \"t\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          elif base_gene == \"t\" and reverse_primer[position_gene] == \"a\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          elif base_gene == \"c\" and reverse_primer[position_gene] == \"g\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          elif base_gene == \"g\" and reverse_primer[position_gene] == \"c\":\n",
    "            position_gene += 1\n",
    "            continue\n",
    "          else: \n",
    "            Primer2 = 1\n",
    "            break\n",
    "            \n",
    "        if Primer1 + Primer2 == 2:\n",
    "          return (\"Both Primers and gene are incompatible\")\n",
    "        elif Primer1 == 1:\n",
    "          return (\"Forward Primer and gene are incompatible\")   \n",
    "        elif Primer2 == 1:\n",
    "          return (\"Reverse Primer and gene are incompatible\")\n",
    "        else:\n",
    "          return (\"Primers and Gene are compatible\")\n",
    "\n",
    "    def iProofAnalyzer(self): \n",
    "        \"\"\"\n",
    "        Acquires iProof DNA Polymerase Results \n",
    "        Based on protocol from iProof: \"https://www.bio-rad.com/webroot/web/pdf/lsr/literature/10002298B.pdf\"\n",
    "\n",
    "        Parameters: \n",
    "        ----------\n",
    "        a pcr object (contains gene seq, f primer, and r primer)\n",
    "\n",
    "        Returns: \n",
    "        ----------\n",
    "        enzyme_df : DataFrame\n",
    "          enzyme volume \n",
    "          enzyme cost\n",
    "        stats : DataFrame \n",
    "          Annealing temperature\n",
    "          Annealing time \n",
    "          Extension time \n",
    "          Total PCR reaction time\n",
    "\n",
    "        \"\"\"\n",
    "        #Gene length and elongation time\n",
    "        #For iProof, extension time depends on the type of template DNA \n",
    "        #Protocol recommends 15 sec per kb if plasmid, and 30 sec per kb if genomic\n",
    "        from Bio.SeqUtils import MeltingTemp as mt\n",
    "        from Bio.Seq import Seq\n",
    "        import pandas as pd\n",
    "\n",
    "        gene_length = len(self.sequence)\n",
    "        if self.template_type == \"plasmid\" or self.template_type == \"lambda\" or self.template_type == \"BAC DNA\": \n",
    "          extension_time_seconds = (gene_length/1000)*15\n",
    "          extension_time_minutes = extension_time_seconds/60 \n",
    "        elif self.template_type == \"genomic\": \n",
    "          extension_time_seconds = round((gene_length/1000)*30, 2)\n",
    "          extension_time_minutes = round(extension_time_seconds/60, 2)\n",
    "\n",
    "        #iProof Enzyme \n",
    "        #iProof can range from 0.5-2 units per 50mL \n",
    "        #Chose the 4 most common PCR volumes. With a higher volume you get more PCR product \n",
    "        \n",
    "        enzyme_dict = {\"20 uL\": [], \"20 uL Cost\" : [], \"50 uL\": [], \"50 uL Cost\" : [], \"100 uL\" : [], \n",
    "                    \"100 uL Cost\": [], \"200 uL\" : [], \"200 uL Cost\" : []} # Store cost for each volume\n",
    "        enzyme_cost = 1.53\n",
    "\n",
    "        for volume in [20,50,100,200]: \n",
    "          enzyme_dict[str(volume) + \" uL\"].append(0.5*(volume/50)) \n",
    "          enzyme_dict[str(volume) + \" uL Cost\"].append(0.5*(volume/50)*enzyme_cost)\n",
    "        enzyme_df = pd.DataFrame(enzyme_dict)\n",
    "        enzyme_df.index = ['Enzyme amount']\n",
    "        \n",
    "        #iProof Primer Annealing Temperature\n",
    "        annealing_temp_1 = mt.Tm_NN(Seq(self.fp))\n",
    "        annealing_temp_2 = mt.Tm_NN(Seq(self.rp))\n",
    "        if len(self.fp) or len(self.fp) > 20: \n",
    "          if annealing_temp_1 > annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_2 + 3, 2)\n",
    "          elif annealing_temp_1 < annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_1 +3, 2)\n",
    "\n",
    "        elif len(self.fp) or len(self.rp) <=20: \n",
    "          if annealing_temp_1 > annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_2,2)\n",
    "          elif annealing_temp_1 < annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_1,2)\n",
    "        annealing_time_seconds = 30 #this is the same regardless of the input\n",
    "        annealing_time_minutes = 0.5 \n",
    "\n",
    "        total_pcr_time_minutes = round((30 + (10 + annealing_time_seconds + extension_time_seconds)*35 + 10*60)/60, 2)\n",
    "        total_pcr_time_hours = round (total_pcr_time_minutes/60, 2)\n",
    "        \n",
    "        factor = ['Annealing Temperature:','Annealing Time:','Extention Time:','Total PCR reaction time is:' ]\n",
    "        stats_data = {'Result': [\"{} degrees Celcius\".format(annealing_temp_final), \"{} seconds or {} minutes\".format(annealing_time_seconds, annealing_time_minutes), \"{} seconds or {} minutes\".format(extension_time_seconds, extension_time_minutes), \"{} minutes or {} hours\".format(total_pcr_time_minutes, total_pcr_time_hours)]}\n",
    "        stats = pd.DataFrame(data = stats_data, index = factor)\n",
    "\n",
    "        return enzyme_df, stats \n",
    "      \n",
    "    def taqAnalyzer (self): \n",
    "        \"\"\"\n",
    "        Acquires Taq DNA Polymerase Results \n",
    "        Based on protocol from Taq: \"https://www.neb.com/protocols/0001/01/01/taq-dna-polymerase-with-standard-taq-buffer-m0273\"\n",
    "\n",
    "        Parameters: \n",
    "        ----------\n",
    "        a pcr object (contains gene seq, f primer, and r primer)\n",
    "\n",
    "        Returns: \n",
    "        ----------\n",
    "        enzyme_df : DataFrame\n",
    "          enzyme volume \n",
    "          enzyme cost\n",
    "        stats : DataFrame \n",
    "          Annealing temperature\n",
    "          Annealing time \n",
    "          Extension time \n",
    "          Total PCR reaction time\n",
    "\n",
    "        \"\"\"\n",
    "        #Gene length and elongation time\n",
    "        #For iProof, extension time depends on the type of template DNA \n",
    "        #Protocol recommends 15 sec per kb if plasmid, and 30 sec per kb if genomic\n",
    "        from Bio.SeqUtils import MeltingTemp as mt\n",
    "        from Bio.Seq import Seq\n",
    "        import pandas as pd\n",
    "\n",
    "        gene_length = len(self.sequence)\n",
    "        if self.template_type == \"plasmid\" or self.template_type == \"lambda\" or self.template_type == \"BAC DNA\": \n",
    "          extension_time_seconds = (gene_length/1000)*15\n",
    "          extension_time_minutes = extension_time_seconds/60 \n",
    "        elif self.template_type == \"genomic\": \n",
    "          extension_time_seconds = round((gene_length/1000)*30, 2)\n",
    "          extension_time_minutes = round(extension_time_seconds/60, 2)\n",
    "\n",
    "        #Taq DNA Polymerase Enzyme \n",
    "        #Taq DNA Polymerase is 0.25 units per 50 \n",
    "        #Chose the 4 most common PCR volumes. With a higher volume you get more PCR product \n",
    "        \n",
    "        enzyme_dict = {\"20 uL\": [], \"20 uL Cost\" : [], \"50 uL\": [], \"50 uL Cost\" : [], \"100 uL\" : [], \n",
    "                    \"100 uL Cost\": [], \"200 uL\" : [], \"200 uL Cost\" : []} # Store cost for each volume\n",
    "        enzyme_cost = 1.78 # dollar cost per reaction\n",
    "        \n",
    "\n",
    "        for volume in [20,50,100,200]: \n",
    "          enzyme_dict[str(volume) + \" uL\"].append(0.25*(volume/50)) \n",
    "          enzyme_dict[str(volume) + \" uL Cost\"].append(0.25*(volume/50)*enzyme_cost)\n",
    "        enzyme_df = pd.DataFrame(enzyme_dict)\n",
    "        enzyme_df.index = ['Enzyme amount']\n",
    "\n",
    "        #iProof Primer Annealing Temperature\n",
    "        annealing_temp_1 = mt.Tm_NN(Seq(self.fp))\n",
    "        annealing_temp_2 = mt.Tm_NN(Seq(self.rp))\n",
    "        if len(self.fp) or len(self.fp) > 20: \n",
    "          if annealing_temp_1 > annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_2 + 3, 2)\n",
    "          elif annealing_temp_1 < annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_1 + 3, 2)\n",
    "\n",
    "        elif len(self.fp) or len(self.rp) <=20: \n",
    "          if annealing_temp_1 > annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_2,2)\n",
    "          elif annealing_temp_1 < annealing_temp_2: \n",
    "            annealing_temp_final = round(annealing_temp_1,2)\n",
    "        annealing_time_seconds = 60 \n",
    "        annealing_time_minutes = annealing_time_seconds/60 \n",
    "\n",
    "        # Time for each Processes in seconds\n",
    "        if test.countDNAbase()[1] < 65:\n",
    "          initial_Denaturation = 30 \n",
    "        else:\n",
    "          initial_Denaturation = 4*60\n",
    "\n",
    "        if annealing_temp_1 > 65 and annealing_temp_2 > 65:\n",
    "          cycles = 30\n",
    "        else:\n",
    "          cycles = 35\n",
    "\n",
    "        Denaturation = 30  \n",
    "        final_extension = 5 * 60 \n",
    "\n",
    "        # Total Time calculation\n",
    "        total_pcr_time_minutes = round((initial_Denaturation + (Denaturation + annealing_time_seconds + extension_time_seconds)*cycles + final_extension)/60, 2)\n",
    "        total_pcr_time_hours = round(total_pcr_time_minutes/60, 2)\n",
    "        \n",
    "        factor = ['Annealing Temperature:','Annealing Time:','Extention Time:','Total PCR reaction time is:' ]\n",
    "        stats_data = {'Result': [\"{} degrees Celcius\".format(annealing_temp_final), \"{} seconds or {} minutes\".format(annealing_time_seconds, annealing_time_minutes), \"{} seconds or {} minutes\".format(extension_time_seconds, extension_time_minutes), \"{} minutes or {} hours\".format(total_pcr_time_minutes, total_pcr_time_hours)]}\n",
    "        stats = pd.DataFrame(data = stats_data, index = factor)\n",
    "\n",
    "        return enzyme_df, stats \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 64-bit (microsoft store)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "55b4ec3b1de4745d4d0bc52a98f917eca26cf669fe460c5ea602bd5130f820d0"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
